### PR TITLE
json: add Amino-compatible encoder/decoder

### PR DESCRIPTION
**Pull request by erikgrinaker**
Amino-compatible JSON encoder/decoder, including bug compatibility. Interface types must be registered via `json.RegisterType()`. Unlike Amino, this allows floats to be encoded/decoded.

Partial fix for #4828, needs code migration.

### CODE REVIEW

File: libs/json/structs.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	"fmt"
+	"reflect"
+	"strings"
+	"unicode"
+
+	lru "github.com/hashicorp/golang-lru"
+)
+
+// structInfoCache caches struct info.
+var structInfoCache *lru.Cache
```
Comment (erikgrinaker) [inline libs/json/structs.go:13]: I'm not really sure if we need this. It caches struct metadata, specifically decoded JSON tags (e.g. `json:"name,omitempy"`), during struct encoding. However, since there is a finite number of structs in the codebase, maybe we should just cache all of them.

File: libs/json/encoder.go [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+func encode(w io.Writer, v interface{}) error {
+	// Bare nil values can't be reflected, so we must handle them here.
+	if v == nil {
+		return writeStr(w, "null")
```
Comment (Liamsi) [inline libs/json/encoder.go:48]: I think it might be important to document either somewhere prominently when `null` is written vs. for instance fields being omitted instead. 

I think this is particularly relevant as some languages (e.g. rust) do not really have a `nil` equivalent (you could argue that `[None](https://doc.rust-lang.org/std/option/enum.Option.html#variant.None)` is similar but it is semantically quite different).

File: libs/json/encoder.go [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+	}
+	rv := reflect.ValueOf(v)
+
+	// If this is a registered type, defer to interface encoder regardless of whether the input is
+	// an interface or a bare value. This retains Amino's behavior, but is inconsistent with
+	// behavior in structs where an interface field will get the type wrapper while a bare value
+	// field will not.
```
Comment (Liamsi) [inline libs/json/encoder.go:55]: Not sure if that actually appears somewhere in tendermint or the SDK but maybe also worth documenting this kind of slight differences in a md document, or even in the changelog / Upgrading.md if it's just a few bulletpoints?

File: libs/json/encoder.go [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+
+func encodeReflect(w io.Writer, rv reflect.Value) error {
+	if !rv.IsValid() {
+		return errors.New("invalid reflect value")
+	}
+
+	// Recursively dereference if pointer.
+	for rv.Kind() == reflect.Ptr {
+		if rv.IsNil() {
+			return writeStr(w, `null`)
```
Comment (Liamsi) [inline libs/json/encoder.go:71]: Why doesn't this use double quotes, like in the `encode` method above? This also might be a boring yet important detail for implementers in other languages.

File: libs/json/encoder.go [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+	// For everything else, defer to the stdlib encoding/json encoder
+	default:
+		return encodeStdlib(w, rv.Interface())
+	}
+}
+
+func encodeReflectList(w io.Writer, rv reflect.Value) error {
+	// Emit nil slices as null.
+	if rv.Kind() == reflect.Slice && rv.IsNil() {
+		return writeStr(w, `null`)
```
Comment (Liamsi) [inline libs/json/encoder.go:121]: Important detail for implementers in other languages.

File: libs/json/encoder.go [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+	for rv.Kind() == reflect.Ptr {
+		if rv.IsNil() {
+			return writeStr(w, `null`)
+		}
+		rv = rv.Elem()
+	}
+
+	// Convert times to UTC.
+	if rv.Type() == timeType {
+		rv = reflect.ValueOf(rv.Interface().(time.Time).Round(0).UTC())
```
Comment (Liamsi) [inline libs/json/encoder.go:78]: The default time in go (`time.Time{}`) used to yield:  value is `0001-01-01T00:00:00Z`. is this still the case? 
This lead to some confusion here for instance:
https://github.com/informalsystems/tendermint-rs/pull/196#discussion_r401027989

Maybe a comment here and a doc.go will provide guidance for implementations in other languages and might prevent bugs there too.

File: libs/json/encoder.go [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+		return encodeReflectList(w, rv)
+
+	case reflect.Map:
+		return encodeReflectMap(w, rv)
+
+	case reflect.Struct:
+		return encodeReflectStruct(w, rv)
+
+	// 64-bit integers are emitted as strings, to avoid precision problems with e.g.
+	// Javascript which uses 64-bit floats (having 53-bit precision).
```
Comment (Liamsi) [inline libs/json/encoder.go:105]: So many good comments ❤️  these should go at least be high lighted in a `doc.go` or maybe even into a markdown document (I'm a fan of package documentation instead of markdown docs because I usually look into code anyways but others prefer specs and markdown docs).

File: libs/json/encoder.go [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+		writeComma = true
+	}
+	return writeStr(w, `}`)
+}
+
+func encodeReflectInterface(w io.Writer, rv reflect.Value) error {
+	// Get concrete value and dereference pointers.
+	for rv.Kind() == reflect.Ptr || rv.Kind() == reflect.Interface {
+		if rv.IsNil() {
+			return writeStr(w, `null`)
```
Comment (Liamsi) [inline libs/json/encoder.go:217]: Nit: Ideally, all cases where `null` are written should be explicitly documented in the package documentation.

File: libs/json/encoder_test.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+				"Car":{"Wheels":4},
+				"Boat":{"Sail":true},
+				"Vehicles":[
+					{"type":"vehicle/car","value":{"Wheels":4}},
+					{"type":"vehicle/boat","value":{"Sail":true}}
+				],
+				"Child":{
+					"Bool":false, "Float64":0, "Int32":0, "Int64":"0", "Int64Ptr":null,
+					"String":"child", "StringPtrPtr":null, "Bytes":null,
+					"Time":"0001-01-01T00:00:00Z",
```
Comment (Liamsi) [inline libs/json/encoder_test.go:86]: Aha, that answers my question about "default" time.

File: libs/json/encoder_test.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+					"Car":null, "Boat":{"Sail":false}, "Vehicles":null, "Child":null
+				}
+			}`,
+		},
+	}
```
Comment (Liamsi) [inline libs/json/encoder_test.go:91]: This really small and quickly written package is at least as well tested as amino if not even better 👍 

Regarding documentation, I have a suggestion: Maybe a very exhaustive [example style test](https://blog.golang.org/examples) inside a `doc.go` would be the best documentation?
It's probably faster to write than well-written and well-structured english documentation and is to the point for everyone looking for answers about edge-cases in documentation.

File: libs/json/structs.go (hunk 1 above)
Comment (Liamsi) [reply libs/json/structs.go:13]: Interesting idea. The whole cache would probably just take a few kb in mem anyways? If that's the case I think it is reasonable to just cache everything 👍

File: libs/json/types.go [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+)
+
+// RegisterType registers a type for Amino-compatible interface encoding in the global type
+// registry. These types will be encoded with a type wrapper `{"type":"<type>","value":<value>}`
+// regardless of which interface they are wrapped in (if any). If the type is a pointer, it will
+// still be valid both for value and pointer types, but decoding into an interface will generate
+// the a value or pointer based on the registered type.
+//
+// Should only be called in init() functions, as it panics on error.
+func RegisterType(_type interface{}, name string) {
```
Comment (Liamsi) [inline libs/json/types.go:22]: On first sight I was wondering if it is bad to "lose" the connection between the registered interfaces and concrete types but the interface doesn't really matter and the connection will still remain visible in the code via the cdc.RegisterXXX methods in the code.

In rust, these registered interfaces + implementations best are represented via a an enum (enum equiv. to interface and enum variants equiv. to concrete impls). 
e.g. https://github.com/informalsystems/tendermint-rs/blob/cf0763e8a5518e35f9155aeee87ce9093fb833e1/tendermint/src/rpc/event_listener.rs#L112-L134

File: libs/json/encoder.go (hunk 6 above)
Comment (erikgrinaker) [reply libs/json/encoder.go:78]: Yes, for backwards compatibility - the idea was to not change anything about RPC until we figure what to do about it long-term. I'll add a `doc.go` with basic encoding details.

File: libs/json/structs.go (hunk 1 above)
Comment (erikgrinaker) [reply libs/json/structs.go:13]: Yeah, I think that makes sense. It'll probably be on the order of MBs, not KBs, but nothing major. I'll do some quick benchmarks to check, and throw out the LRU.

File: libs/json/encoder.go (hunk 4 above)
Comment (erikgrinaker) [reply libs/json/encoder.go:71]: This is just Go string quoting, where backticks are useful for strings that may contain quotes. No reason to use it here, will switch to `"`.

File: libs/json/encoder_test.go (hunk 10 above)
Comment (erikgrinaker) [reply libs/json/encoder_test.go:91]: Good idea.

File: libs/json/types.go [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+)
+
+var (
+	// typeRegistry contains globally registered types for JSON encoding/decoding.
+	typeRegistry = newTypes()
```
Comment (Liamsi) [inline libs/json/types.go:12]: Another note: As far I remember amino put in a non-trivial and confusing amount of guarding the codec via mutexes. I highly doubt that anyone was ever using it concurrently (I mean ... why on earth would you spin up some go-routines to compete with registering your types etc 😬 ). Nonetheless, this is a change in behaviour that one _could_ mention in an upgrading doc (so users that want to switch to using this library are aware that this isn't go-routine safe anymore)?

File: libs/json/types.go (hunk 11 above)
Comment (erikgrinaker) [reply libs/json/types.go:22]: For JSON we never really use the interface for anything, and I'd ideally like to strip this down to the absolute minimum. I think there's only a handful of registered types which will remain (mainly the `Evidence` interface and encryption keys), hopefully we can throw this entire package out when we clean up the RPC layer.

File: libs/json/types.go (hunk 12 above)
Comment (erikgrinaker) [reply libs/json/types.go:12]: This is goroutine-safe, the type registry has an internal `RWMutex`. And the `typeRegistry` variable will be initialized when the package is loaded, the same package that contains `RegisterType()`. I think we're good?

The only change, I suppose, is that there is now a common mutex across all JSON encoders, while with Amino one could in principle have multiple codecs with separate mutexes. However, from the uses I've come across there's always a single codec.

Comment (Liamsi) [reply libs/json/types.go:12]: Ah, I didn't see the `sync.RWMutex`. So it isn't even sth to mention. Sorry for the noise.

File: libs/json/decoder.go [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+			rv.Set(reflect.New(rv.Type().Elem()))
+		}
+		rv = rv.Elem()
+	}
+
+	// Times must be UTC and end with Z
+	if rv.Type() == timeType {
+		switch {
+		case len(bz) < 2 || bz[0] != '"' || bz[len(bz)-1] != '"':
+			return fmt.Errorf("JSON time must be an RFC3339 string, but got %s", bz)
```
Comment (melekes) [inline libs/json/decoder.go:62]:
```suggestion
			return fmt.Errorf("JSON time must be an RFC3339 string, but got %q", bz)
```

File: libs/json/decoder.go [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+		case bz[len(bz)-2] != 'Z':
+			return fmt.Errorf("JSON time must be UTC and end with 'Z', but got %s", bz)
```
Comment (melekes) [inline libs/json/decoder.go:64]:
```suggestion
			return fmt.Errorf("JSON time must be UTC and end with 'Z', but got %q", bz)
```

File: libs/json/decoder.go [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+		}
+		rv = rv.Elem()
+	}
+
+	// Times must be UTC and end with Z
+	if rv.Type() == timeType {
+		switch {
+		case len(bz) < 2 || bz[0] != '"' || bz[len(bz)-1] != '"':
+			return fmt.Errorf("JSON time must be an RFC3339 string, but got %s", bz)
+		case bz[len(bz)-2] != 'Z':
```
Comment (melekes) [inline libs/json/decoder.go:63]: don't we want to check both cases here?

Comment (erikgrinaker) [reply libs/json/decoder.go:63]: What other case?

Comment (erikgrinaker) [reply libs/json/decoder.go:63]: Not sure I follow. Both branches are checked? Do you mean combining them into a single check with a common error message?

Comment (melekes) [reply libs/json/decoder.go:63]: sorry, it's just a confusing style of checking for errors...

i'd much prefer 

if X {
  return E
}

if Y {
  return E
}

File: libs/json/encoder_test.go (hunk 10 above)
Comment (erikgrinaker) [reply libs/json/encoder_test.go:91]: Ok, so I tried the example test approach, but it's a bit tricky since it only allows a _single_ output for a test. So this won't work, because only the last `Output` takes effect:

```go
func Example() {
	printJSON("foo")     // Output: "foo"
	printJSON(int64(64)) // Output: "64"
	printJSON(int32(32)) // Output: 32
}
```

```
=== RUN   Example
--- FAIL: Example (0.00s)
got:
"foo"
"64"
32
want:
32
FAIL
FAIL	github.com/tendermint/tendermint/libs/json	0.241s
FAIL
```

I can add one example test per case:

```
func ExampleMarshal_i32()    { printJSON(int32(32)) /* Output: 32 */ }
func ExampleMarshal_i64()    { printJSON(int64(64)) /* Output: "64" */ }
func ExampleMarshal_string() { printJSON("foo") /* Output: "foo" */ }
```

But this results in very verbose documentation, with 4 paragraphs of doc output for every single test.

I'll just add a doc comment instead.

Comment (Liamsi) [reply libs/json/encoder_test.go:91]: I see 👍 

An alternative approach could be to group a bunch of basic types into an array and (pretty)print them into a single output. 

The edge cases (`null`, `Time` etc) could deserve separate paragraphs even. But a doc comment is probably as good as this. I only suggested  Example tests because I thought they'll be faster to write actually (yet contain all relevant info).

File: libs/json/encoder.go (hunk 8 above)
Comment (erikgrinaker) [reply libs/json/encoder.go]: This really depends on what we're encoding though, it's up to the caller whether an input is a pointer or a bare value (the first will emit `null`, the latter will emit the zero value).

File: libs/json/encoder_test.go (hunk 10 above)
Comment (erikgrinaker) [reply libs/json/encoder_test.go:95]: I've added package-level documentation, let me know if you think anything is lacking. Will merge after lunch.

File: libs/json/encoder_test.go [hunk 16]
```diff
@@ -41,15 +41,19 @@ func TestMarshal(t *testing.T) {
...
 		"customvalue ptr": {&CustomValue{Value: "x"}, `"custom"`},
 		"slice nil":       {[]int(nil), `null`},
+		"slice empty":     {[]int{}, `[]`},
 		"slice bytes":     {[]byte{1, 2, 3}, `"AQID"`},
 		"slice int64":     {[]int64{1, 2, 3}, `["1","2","3"]`},
 		"slice int64 ptr": {[]*int64{&i64, nil}, `["64",null]`},
 		"array bytes":     {[3]byte{1, 2, 3}, `"AQID"`},
 		"array int64":     {[3]int64{1, 2, 3}, `["1","2","3"]`},
+		"map nil":         {map[string]int64(nil), `{}`}, // retain Amino compatibility
+		"map empty":       {map[string]int64{}, `{}`},
```
Comment (Liamsi) [inline libs/json/encoder_test.go:51]: 👍

File: libs/json/encoder_test.go (hunk 10 above)
Comment (Liamsi) [reply libs/json/encoder_test.go:95]: LGTM 👍
